### ISSUE TITLE
Make OrganizeImportsService public

### ISSUE THREAD

**Issue by tiesmaster** [dotnet/roslyn#20988]
I'd like to invoke [`OrganizeImportsService.OrganizeImportsAsync()`](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/OrganizeImports/OrganizeImportsService.cs,ed8f2be1e4071c27) from a code refactoring, however, it's  not public.

Now, I need to resort to [using reflection](https://github.com/tiesmaster/AdhocAnalyzers/blob/057f40bfa0b634803b32edf42b9a64027f13fce3/AdhocRefactorings/AdhocRefactorings/StructureNamespaceUsingsRefactoringProvider.cs#L25-L29) to invoke this service, but it would make the code a lot cleaner to do this using the "front door".

**Comment by CyrusNajmabadi** [comment]
What is your full scenario?  Why do you need to call this?

**Comment by CyrusNajmabadi** [comment]
Ah i see.  We might be able to add something to ImportAdder to help out here.

**Comment by Pilchie** [comment]
Also tagging @DustinCampbell on this.

**Comment by tiesmaster** [comment]
# TL;DR version of this
I'd like to be able to have my refactoring have two code actions: Add newlines between usings (top level) groups, and one where it also does that, but first runs "Sort, and remove unused usings".

# Background
I've been a ReSharper addict for many, many years, but with everything people are addicted to, there are downsides (when I moved to VS 2015, performance was basically out of the window, also due to having other plugins installed, like NCrunch, which doesn't help). I've tried de-installing R# many times, but the gap was always too big. However, VS 2017, of course, added a ton of good stuff, I really wanted to take the effort to come clean and start working without R#.

However, having a rule in our companies coding standard to actually run R# code cleanup with a certain custom profile, I have a dependency on that tool to get my code formatted following a number of rules. One of the rules it has, is adding newlines between using groups, and I wanted to make a refactoring that actually provides me with that, so I can easily apply that rule through a Roslyn refactoring.

I've done that, and that's working great, but I always invoke "Sort, and remove unused usings", and then I invoke my refactoring. I really like that to be one action, to make it easier. Having `OrganizeImportsService` public would make it possible to at least organize the usings. For removing the unused usings, it would be handy to have `IRemoveUnnecessaryImportsService` be public as well, but I could handle that differently (I believe unused usings are also surfaced as "hidden" warnings, and handle that from there).

# Crazy Idea Section
It would be even more awesome if you could replace the `IOrganizeImportsService` with a different implementation at runtime, so invoking "Sort, and remove unused usings" would invoke a custom implementation of the organize import service. This is not possible, since that interface is also not public, but I don't even know if it's possible to replace a language service through an extension, since I haven't found any public `ILanguageService` to experiment, and validate that idea.

In the same vein, it would be awesome to be able to make a plugin that would replace the `IFormattingService` with your own implementation, such as wrapping [dotnet/codeformatter](https://github.com/dotnet/codeformatter) (or something similar) to format your code (similar as R#'s cleanup code would do).

That's basically my end goal, I don't know if that makes any sense, or is totally absurd. I know that this is not what the issue is about, and I can open a new issue for this, if needed. But first, I'd like to hear if this makes any sense, or these train of thoughts are going nowhere ;)

**Comment by CyrusNajmabadi** [comment]
> Add newlines between usings (top level) groups, and one where it also does that, but first runs "Sort, and remove unused usings".

@tiesmaster Why not just have roslyn include such an option in our default option set/.editor_config support?  I'd be happy to include such an option.  @kuhlenh @DustinCampbell  what do you think?

**Comment by tiesmaster** [comment]
@CyrusNajmabadi That's awesome!! I didn't even think of requesting such a feature, as I thought it would be too specific. Nice that you could already implement it 😄  and put it in a PR. Any idea if this will make it to 15.3, or would this be scheduled much later, if merged?

**Comment by CyrusNajmabadi** [comment]
Definitely could not make 15.3.  That branch is locked down for anything that's not something critical (i.e. 'bad') that is hitting users.

**Comment by tiesmaster** [comment]
Btw does my "Crazy Idea"-section make any sense? In the mean time, I've noticed the projects section in this repo for specifically supporting [more refactorings](https://github.com/dotnet/roslyn/projects/5), and [code style](https://github.com/dotnet/roslyn/projects/4). So I can imagine my mentioned approach is not in line with the vision of Roslyn/Visual Studio, but I think you guys are striving for something similar, isn't it?

**Comment by CyrusNajmabadi** [comment]
Yes.  it makes sense.  The difficulty comes in with designing how that would work and providing a sensible user experience around it.  A primary issue is simply that once we create one of these systems, we effectively cannot change it as that would break users.  That means we really need to get things *right* from the beginning.
